dispatch/dispatch.go

```diff
@@ -517,19 +517,11 @@ func (ag *aggrGroup) flush(notify func(...*types.Alert) bool) {
 
 	if notify(alertsSlice...) {
 		for _, a := range alertsSlice {
-			// Only delete if the fingerprint has not been inserted
```

Comment:
[nit] How do you feel about making the whole operation atomic? You'd only take the lock once guaranteed that all the alerts are cleared in one go. I feel this makes it easier to reason about it as there's no use case where an alert could get updated once we start checking. 

```git
diff --git a/dispatch/dispatch.go b/dispatch/dispatch.go
index d78fbf68..6e44fbc2 100644
--- a/dispatch/dispatch.go
+++ b/dispatch/dispatch.go
@@ -516,14 +516,7 @@ func (ag *aggrGroup) flush(notify func(...*types.Alert) bool) {
 	level.Debug(ag.logger).Log("msg", "flushing", "alerts", fmt.Sprintf("%v", alertsSlice))
 
 	if notify(alertsSlice...) {
-		for _, a := range alertsSlice {
-			// Only delete if the fingerprint has not been inserted again since we notified about it.
-			if err := ag.alerts.DeleteIf(a.Fingerprint(), func(b *types.Alert) bool {
-				return a.Resolved() && a.UpdatedAt == b.UpdatedAt
-			}); err != nil {
-				level.Error(ag.logger).Log("msg", "error on delete alert", "err", err, "alert", a.String())
-			}
-		}
+		ag.alerts.DeleteResolved(alertsSlice)
 	}
 }
 
diff --git a/store/store.go b/store/store.go
index bffa4eb6..4110a971 100644
--- a/store/store.go
+++ b/store/store.go
@@ -154,3 +154,19 @@ func (a *Alerts) Empty() bool {
 
 	return len(a.c) == 0
 }
+
+func (a *Alerts) DeleteResolved(slice types.AlertSlice) {
+	a.Lock()
+	defer a.Unlock()
+
+	for _, as := range slice {
+		fp := as.Fingerprint()
+		alert, ok := a.c[as.Fingerprint()]
+		if !ok {
+		}
+
+		if as.Resolved() && as.UpdatedAt == alert.UpdatedAt {
+			delete(a.c, fp)
+		}
+	}
+}

```

Reply:
I quite like the condition being a closure, so what if we had something like:

```go
func (a *Alerts) DeleteFingerprints(fp []model.Fingerprint, fn func(int, *types.Alert) bool) error {
	...
}
```

I did think about adding something like `DeleteResolved` but with fingerprints:

```go
func (a *Alerts) DeleteResolved(fp []model.Fingerprint) error {
	...
}
```

but I decided I didn't like the semantics of it because it brought business logic from the dispatcher [1] into the store, and I felt the store should be agnostic of that kind of business logic.

[1] For example:

1. The name `DeleteResolved` felt confusing as you would expect a store with a method called `DeleteResolved` to delete all resolved alerts it knows about, not the resolved alerts the caller knows about (it doesn't delete all resolved alerts, just those in the argument).
2. If the caller thinks the alert is firing, but the store thinks the alert is resolved, then `DeleteResolved` will not delete the alert. This feels weird given the method is called `DeleteResolved`.
3. A resolved alert is deleted iff UpdatedAt hasn't changed since the last flush (but why does a generic store care about flushes?).

What do you think?

Reply:
That's a fair point. 

I'm OK with:

> func (a *Alerts) DeleteFingerprints(fp []model.Fingerprint, fn func(int, *types.Alert) bool) error {

As what I care about is the atomicity of the operation. 

I think your point about logic is valid but they both share the same types anyways 🤷 so I kind feel that where the logic goes doesn't matter to me as much.